File: QiMata.Datasheet.AmpWordParser/Services/ComparisonService.cpp
#include <stdafx.h>
#include "ComparisonService.hpp"

using namespace QiMata::Datasheet::AmpWordParser::Models;
using namespace QiMata::Datasheet::AmpWordParser::Services;

std::vector<StringSectionMatch> ComparisonService::ComapreStringSections(FlattenedDatasheet datasheet)
{
	std::vector<StringSectionMatch> matches(datasheet.FlattenedSections.size() * datasheet.FlattenedSections[0].StringSections.size());
	
	for (auto& section : datasheet.FlattenedSections)
	{

		//TODO: make parallel
		for (auto& x : section.StringSections)
		{
			//TODO: make parallel
			for (auto& y : section.StringSections)
			{
				if (x.Id != y.Id)
				{
					if (x.StringHash == y.StringHash && x.StringLength == y.StringLength)
					{
						matches.push_back(StringSectionMatch{ x, y });
					}
				}
			}
		}
	}

	return matches;
}

File: QiMata.Datasheet.AmpPassThrough/Services/ComparisonService.cs
using QiMata.Datasheet.AmpPassThrough.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace QiMata.Datasheet.AmpPassThrough.Services
{
    public class ComparisonService : IComparisonService
    {
        public IEnumerable<StringSectionMatch> ComapreStringSections(FlattenedDatasheet datasheet1, FlattenedDatasheet datasheet2)
        {
            List<StringSectionMatch> matches = new List<StringSectionMatch>(datasheet1.FlattenedSections.Count * datasheet2.FlattenedSections.Count);
            Parallel.ForEach(datasheet1.FlattenedSections, x =>
                {
                    Parallel.ForEach(datasheet2.FlattenedSections, y =>
                        {
                            foreach(var section1 in x.StringSections)
                            {
                                foreach(var section2 in y.StringSections)
                                {
                                    if (section1.Id != section2.Id)
                                    {
                                        if (section1.StringHash == section2.StringHash && section1.StringLength == section2.StringLength)
                                        {
                                            matches.Add(new StringSectionMatch
                                            {
                                                Section1 = section1,
                                                Section2 = section2
                                            });
                                        }
                                    }
                                }
                            }
                        });
                });
            return matches; 
        }
    }
}
